### PR TITLE
mac-virtualcam: Avoid transcoding where possible

### LINKED ISSUE #6285: mac-virtualcam DAL plugin has incompatible architectures
### Operating System Info

macOS 12.0

### Other OS

_No response_

### OBS Studio Version

27.2.4

### OBS Studio Version (Other)

_No response_

### OBS Studio Log URL

https://obsproject.com/logs/J07TkJBDIfo6Igsn

### OBS Studio Crash Log URL

_No response_

### Expected Behavior

Virtualcam showing on chrome. Because I do see Chrome on the Compatibility list of virtualcam.
(https://github.com/johnboiles/obs-mac-virtualcam/wiki/Compatibility)

### Current Behavior

Only facetime HD showing on Chrome. 
image

**AND virtualcam also NOT showing on any other apps**

### Steps to Reproduce

1. I check _/Library/CoreMediaIO/Plug-Ins/DAL_ and _/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/MacOS/_ , it's there.
image
image

2. I check **system_profiler SPCameraDataType**, and here is what I got
```
vickyching@VickydeAir ~ % system_profiler SPCameraDataType
2022-04-06 16:30:17.989 system_profiler[94171:23973235] Error loading /Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/MacOS/obs-mac-virtualcam:  dlopen(/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/MacOS/obs-mac-virtualcam, 0x0106): tried: '/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/MacOS/obs-mac-virtualcam' (fat file, but missing compatible architecture (have 'x86_64,arm64', need 'arm64e'))
Camera:

    FaceTime HD Camera:

      Model ID: FaceTime HD Camera
      Unique ID: EAB7A68FEC2B4487AADFD8A91C1CB782
```

### Anything else we should know?

_No response_

### PR DESCRIPTION
**Pull request by fabianishere**
This pull request updates the `mac-virtualcam` plugin to avoid transcoding the video feed if possible. 

Additionally, I have updated the DAL plugin to build for `ARM64e` as well (trying to fix #6285)

### Motivation and Context
Currently, the `mac-virtualcam` plugin transcodes the video feed into `UYVY`, since the DAL plugin exposes that format to the consuming application. However, having to do this transformation incurs higher CPU usage (see https://github.com/johnboiles/obs-mac-virtualcam/issues/102 for more information).

This pull request updates the plugin to support transferring the raw video feed to the DAL plugin without transformation for the pixel formats most commonly used in OBS (NV12 and I444, I believe). We achieve this by communicating with `CVPixelBuffer`s between the OBS plugin and the DAL plugin, as opposed to raw data buffers. `CVPixelBuffer` objects carry additional information (such as pixel format and video dimensions) and can be converted quite easily to `CMSampleBuffer` objects that are submitted to the consuming application.

### How Has This Been Tested?
These changes have been tested on a Macbook Pro (M1) running macOS Monterey and have been verified to work with Google Chrome and Zoom.

### Types of changes
- Performance enhancement: transcoding the video feed is avoided if possible. Pixel buffers are properly shared across processes (zero-copy, via IOSurface). 
- Code cleanup: removal of unused functions.
- (Potentially) breaking change: communication protocol between OBS plugin and DAL plugin has changed. If user is still using the old DAL plugin, the virtual cam will not work.

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm [hunk 1]
```diff
@@ -163,33 +212,98 @@ static bool virtualcam_output_start(void *data)
...
-	if (frame->linesize[0] != (videoInfo.output_width * 2)) {
-		blog(LOG_ERROR,
-		     "unexpected frame->linesize (expected:%d actual:%d)",
-		     (videoInfo.output_width * 2), frame->linesize[0]);
+	CVPixelBufferRef frameRef = nil;
+	CVReturn status = CVPixelBufferPoolCreatePixelBuffer(
+		kCFAllocatorDefault, vcam->pool, &frameRef);
+
+	if (status != kCVReturnSuccess) {
+		blog(LOG_ERROR, "unable to allocate pixel buffer (error %d)",
```
Comment (gxalpha) [inline plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:235]: See above

**Review by gxalpha — CHANGES_REQUESTED** [review]
Some initial impressions, didn’t look too deeply into the code or build or test yet.

Regarding your comments about this being breaking: Changing the communication protocol is completely fine, OBS will automatically update the DAL plugin when a new version comes out, running the old one isn’t a supported use case.

File: plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm [hunk 2]
```diff
@@ -148,22 +142,25 @@ - (void)sendFrameWithSize:(NSSize)size
...
-		NSData *frameData = [NSData
-			dataWithBytesNoCopy:(void *)frameBytes
-				     length:size.width * size.height * 2
-			       freeWhenDone:NO];
+		mach_port_t framePort = IOSurfaceCreateMachPort(
+			CVPixelBufferGetIOSurface(frame));
+
+		if (!framePort) {
+			blog(LOG_ERROR,
+			     "unable to allocate mach port for pixel buffer");
```
[the remark this answers is not in the document]
Comment (fabianishere) [reply plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm:150]: The `blog` function is actually redefined in `Defines.h`:
https://github.com/obsproject/obs-studio/blob/34e346a011549053d4da2b2a52c29de501e9ea73/plugins/mac-virtualcam/src/obs-plugin/Defines.h#L20-L23 

This ensures that the prefix is always added to the message.

**Comment by fabianishere** [comment]
@gxalpha Thanks for the comments. I have updated the pull request according to your suggestions.

File: plugins/mac-virtualcam/src/dal-plugin/CMakeLists.txt [hunk 3]
```diff
@@ -58,6 +60,7 @@ set_target_properties(
              FOLDER "plugins"
              VERSION "0"
              SOVERSION "0"
+             OSX_ARCHITECTURES "x86_64;arm64;arm64e"
```
[the remark this answers is not in the document]
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/dal-plugin/CMakeLists.txt:62]: I don't think the change would be propagated upwards in the build hierarchy, but I'm fine with it as is.

`arm64e` is actually the more "correct" architecture for M1, but as mentioned it was tough enough to get some dependencies to build for arm64 to begin with.

**Comment by PatTheMav** [comment]
Looks fine to me overall - using a `CVPixelBuffer` brings it closer to how Apple's own virtual camera extensions work, so seems like the more "correct" way to achieve this.

File: plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm (hunk 2 above)
[the remark this answers is not in the document]
Comment (gxalpha) [reply plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm:150]: Ah, I forgot about that part. Nevermind.

**Review by gxalpha — APPROVED** [review]

**Comment by PatTheMav** [comment]
I did some tests on my local machines:

* On my M1-based Mac mini this worked without issue
* On my x86_64-based iMac the virtual camera output was either stuck on a single-coloured frame or wasn't updated at all.

Both Macs run macOS 12.4.

**Comment by PatTheMav** [comment]
Seems I spoke to soon, it seized to work on my M1 Mac as well - will have to investigate..

**Comment by fabianishere** [comment]
@PatTheMav How did you test the virtual cam? Hopefully I can reproduce your issue as well.

**Comment by PatTheMav** [comment]
> @PatTheMav How did you test the virtual cam? Hopefully I can reproduce your issue as well.

Nothing special, pulled the code, built OBS, activated the cam, had it overwrite the preexisting one. Tested with https://github.com/lvsti/Cameo.

I got crashes in Cameo as well as OBS now, both of which I have to further investigate right now.

**Comment by fabianishere** [comment]
@PatTheMav Looks like I am able to reproduce the issue with Cameo. Will investigate.

**Comment by fabianishere** [comment]
@PatTheMav I think I found the cause of the issues (which the updated pull request should fix):

1.  **Virtual camera output stuck on a single-coloured frame or not updated at all** 
     I originally did not use `CMIOSampleBufferCreateForImageBuffer` to create the `CMSampleBuffer`. Apparently, this is fine for some applications (e.g., Zoom and Chrome), but not for others (e.g., Cameo).
2.  **OBS or Application crashing**  
     It seems the original code tried to double-free the Mach port that was used to exchange the IOSurfaces.

**Comment by PatTheMav** [comment]
> @PatTheMav I think I found the cause of the issues (which the updated pull request should fix):
> 
> 1. **Virtual camera output stuck on a single-coloured frame or not updated at all**
>    I originally did not use `CMIOSampleBufferCreateForImageBuffer` to create the `CMSampleBuffer`. Apparently, this is fine for some applications (e.g., Zoom and Chrome), but not for others (e.g., Cameo).
> 2. **OBS or Application crashing**
>    It seems the original code tried to double-free the Mach port that was used to exchange the IOSurfaces.
> 
> Cameo now works on my device as well (where it did not before).

Great, seems to work in my tests on both architectures now as well.

**Comment by PatTheMav** [comment]
@gxalpha did you test this already?

**Review by PatTheMav — APPROVED** [review]
Went through the code again, looks fine to me, changes are straight-forward enough.

At some point we should move to a HelperTool + XPCService setup, probably around the time we add a modern VirtualCamera output, but should be fine for now.

**Comment by fabianishere** [comment]
@PatTheMav I have already worked a bit on a prototype using XPC as well as one using the new [CMIO System Extension framework](https://developer.apple.com/documentation/coremediaio/creating_a_camera_extension_with_core_media_i_o) (in Swift). I could create a (draft) pull request once those changes are ready for integration.

Using the new CMIO framework vastly simplifies the implementation of the virtual camera, but the development experience is quite painful, since System Extensions require special entitlements and other strict requirements (e.g., application must be located in `/Applications`)

**Comment by PatTheMav** [comment]
@fabianishere that'd be amazing, @gxalpha and I have been looking at that precisely and we'd be very happy to check out what you did so far. 👍🏻

**Comment by fabianishere** [comment]
I can only reproduce this with Zoom, not with Chrome or Cameo. It looks like Zoom is somehow not releasing the `IOSurfaces` to be re-used by OBS Studio, causing it to allocate more and more memory.

**Comment by SciTechNick** [comment]
FYI: I've just confirmed the same issue in Microsoft Edge + Teams Progressive Web App in a video call.

**Comment by fabianishere** [comment]
Looks like OBS is leaking Mach ports and we need to manually deallocate the ports using `mach_port_deallocate`. The following patch fixes the memory leak for me:

```diff
diff --git a/plugins/mac-virtualcam/src/dal-plugin/OBSDALMachClient.mm b/plugins/mac-virtualcam/src/dal-plugin/OBSDALMachClient.mm
index 452ee6fbc..f781b29f5 100644
--- a/plugins/mac-virtualcam/src/dal-plugin/OBSDALMachClient.mm
+++ b/plugins/mac-virtualcam/src/dal-plugin/OBSDALMachClient.mm
@@ -101,14 +101,32 @@
 		break;
 	case MachMsgIdFrame:
 		VLog(@"Received frame message");
-		if (components.count >= 4) {
+
+		if (components.count < 4) {
+			break;
+		}
+
+		@autoreleasepool {
 			NSMachPort *framePort = (NSMachPort *)components[0];
+
+			if (!framePort) {
+				return;
+			}
+
 			IOSurfaceRef surface = IOSurfaceLookupFromMachPort(
 				[framePort machPort]);
+			mach_port_deallocate(mach_task_self(),
+					     [framePort machPort]);
+
+			if (!surface) {
+				ELog(@"Failed to obtain IOSurface from Mach port");
+				return;
+			}
 
 			CVPixelBufferRef frame;
 			CVPixelBufferCreateWithIOSurface(kCFAllocatorDefault,
 							 surface, NULL, &frame);
+			CFRelease(surface);
 
 			uint64_t timestamp;
 			[components[1] getBytes:&timestamp
@@ -131,7 +149,6 @@
 					    fpsDenominator:fpsDenominator];
 
 			CVPixelBufferRelease(frame);
-			CFRelease(surface);
 		}
 		break;
 	case MachMsgIdStop:
diff --git a/plugins/mac-virtualcam/src/dal-plugin/OBSDALStream.mm b/plugins/mac-virtualcam/src/dal-plugin/OBSDALStream.mm
index b318f6fb7..571eefd70 100644
--- a/plugins/mac-virtualcam/src/dal-plugin/OBSDALStream.mm
+++ b/plugins/mac-virtualcam/src/dal-plugin/OBSDALStream.mm
@@ -397,7 +397,10 @@
 	}
 
 	err = CMSimpleQueueEnqueue(self.queue, sampleBuffer);
+
 	if (err != noErr) {
+		CFRelease(sampleBuffer);
+
 		DLog(@"CMSimpleQueueEnqueue err %d", err);
 		return;
 	}
diff --git a/plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm b/plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm
index d2857766e..efd33c1b4 100644
--- a/plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm
+++ b/plugins/mac-virtualcam/src/obs-plugin/OBSDALMachServer.mm
@@ -147,23 +147,33 @@
 			dataWithBytes:&fpsDenominator
 			       length:sizeof(fpsDenominator)];
 
-		NSPort *framePort = [NSMachPort
-			portWithMachPort:IOSurfaceCreateMachPort(
-						 CVPixelBufferGetIOSurface(
-							 frame))];
+		IOSurfaceRef surface = CVPixelBufferGetIOSurface(frame);
+
+		if (!surface) {
+			blog(LOG_ERROR,
+			     "unable to access IOSurface associated with CVPixelBuffer");
+			return;
+		}
+
+		mach_port_t framePort = IOSurfaceCreateMachPort(surface);
 
 		if (!framePort) {
 			blog(LOG_ERROR,
-			     "unable to allocate mach port for pixel buffer");
+			     "unable to allocate mach port for IOSurface");
 			return;
 		}
 
 		[self sendMessageToClientsWithMsgId:MachMsgIdFrame
 					 components:@[
-						 framePort, timestampData,
+						 [NSMachPort
+							 portWithMachPort:framePort
+								  options:NSMachPortDeallocateNone],
+						 timestampData,
 						 fpsNumeratorData,
 						 fpsDenominatorData
 					 ]];
+
+		mach_port_deallocate(mach_task_self(), framePort);
 	}
 }
 ```

**Comment by gxalpha** [comment]
@fabianishere Thanks for the quick fix, would you be interested in submitting another PR? Otherwise we could test and submit it as well.

**Pull request merged**